Improved cif parsing checks

mbercx opened the pull request:
Fixes #98

From @lorisercole's original description in #106

---

Detects the wrong parsing of certain CIF files with missing sites (fixes #98).

### Chemical formula checks
#### check_formula work function
The new work function `check_formula` compares the formula reported in the CIF file with the composition of a `StructureData`.
It raises specific exceptions if:
- the structure has *missing/additional elements*, compared to the CIF formula
- the *composition is different* to the one reported in the CIF formula

Vacancies (element `'X'`) are ignored.

In order to compare the compositions, the ratios between the amounts of the elements are compared (for each element: `ratio = element_count_CIF_formula / element_count_StructureData_formula`). All the ratios should be the same. 
There is a *tolerance* that should account for the finite precision of the amounts reported in the chemical formula.
- For partially-occupied structures, a good threshold is an absolute tolerance of 0.05 + a 2% relative tolerance. Of course in general this threshold might be too strict or too loose. Often the chemical formulas reported are approximated.
- For structures with no partial occupancies, the threshold is the minimum ratio difference that would produce a StructureData with one atom more/less.

#### cif-clean workchain
I added a step `check_chemical_formula` to the cif-clean workchain. It is executed *after* the `parse_cif_structure` step, and can be skipped using the `skip_formula_check` input flag. If the check fails, the workchain fails, i.e. it does not return a StructureData (even though this node was generated in the previous step).

### Other additions
- `occupancy_tolerance` added to the inputs of the cif-clean workchain. It is used in the `primitive_structure_from_cif` calcfunction. I noticed that sometimes the site occupations of CIF files with partial occupancies are not super precise (few digits reported). This may cause the `cif.get_structure` method to fail because site occupations can be slightly above 1.0. A small occupancy tolerance (above 1.0) helps to avoid this issue when needed.
- `sym_prec` and `site_tolerance` were hard-coded in the `cif_clean` CLI: I added them to the inputs.
- possibility of manually selecting multiple nodes in the `cif_clean` CLI
- option to query for specific elements when importing CIF files using the `aiida-codtools` CLI.

lorisercole committed:
Option to query for elements during cif import

With the option -e ELEMENT it is now possible to import cif structures
with specific elements. Repeat the option for multiple elements:
aiida-codtools data cif import -d cod -G 10 -e Li -e O

lorisercole committed:
Add formula check to cif cleaner

Vacancies will be ignored in missing-elements formula check

Add chemical formula check to CifCleanWorkChain

A new step `check_chemical_formula` has been added to the `CifCleanWorkChain`.
The composition of the `StructuraData` created by `primitive_structure_from_cif`
in the previous step is checked against the chemical formula reported in
the `CifData` node. If it is compatible, the structure will be used as
output; otherwise an error will be raised and no structure will be
returned by the workchain. This check can be skipped by setting the
`skip_formula_check` input variable to True.

mbercx committed:
Add extra mpds tag to `get_formula_from_cif`

The previous version of `get_formula_from_cif` would only check for the
`_pauling_file_chemical_formula`. However, there is also the
`_pauling_file_phase` tag that contains the formula in the first part of
a sequence separated by `/`.

Here we also look for this tag after parsing the `.cif`, and extract the
corresponding formula.

Finally, there is also a small refactoring of the code using the
`get_el_amt_dict` of the pymatgen `Composition` class.

mbercx commented:
Most of the credit here goes to @lorisercole, I just made some minor changes so far to also include a different MPDS tag and clean up the code a bit.

@sphuber I'll have a closer look again at the code to see if I would still make any other changes, feel tree to already add comments as well.

lorisercole commented:
Thanks @mbercx 
Don't know what happened, but it looks like there are some merge conflicts that have not been resolved...

sphuber commented:
There are a bunch of errors from rebasing and the files are not added in the `src` directory. I will fix it

sphuber committed:
Remove adding of `has_partial_occupancies` extra to `StructureData`

This extra is set in `primitive_structure_from_cif` using a function
that is a one-liner. It is then used in `check_formula`. It is better to
just run this one line of code directly in `check_formula` because in
this way it functioning properly is not dependent on the structure
having been generated by `primitive_structure_from_cif` but any
structure can be used.

sphuber committed:
Use `parse_formula` of `aiida.orm.nodes.data.cif`

sphuber committed:
Fix changes of first commit

The default for `element` is `()` and not `None`.

mbercx commented:
Great, thanks @sphuber! Will refresh the code again in my mind and see if we should still make any changes. I'll also contact @lorisercole so we can schedule a meeting for next week.

lorisercole commented on src/aiida_codtools/calculations/functions/check_formula.py:
```diff
@@ -0,0 +1,143 @@
+        # for integer occ., compare ratios against the first one rounded to the nearest integer
+        # use only the absolute tolerance estimated above
+        compare_with_ratio = np.round(ratios[0])
+        rtol = 0.
```
For _integer_ occupancies, I set the _relative_ tolerance to 0 and use only an absolute one.
The absolute tolerance is chosen to be the largest one possible: if the difference in ratios is larger than this tolerance, it means that there is at least a one-atom difference in the formulas.

lorisercole commented on src/aiida_codtools/calculations/functions/check_formula.py:
```diff
@@ -0,0 +1,143 @@
+        compare_with_ratio = np.round(ratios[0])
+        rtol = 0.
+
+    bad_ratios = not np.allclose(ratios, compare_with_ratio, atol=atol, rtol=rtol)
```
From [numpy.allclose](https://numpy.org/doc/stable/reference/generated/numpy.allclose.html) documentation:

> If the following equation is element-wise True, then allclose returns True.
> `absolute(a - b) <= (atol + rtol * absolute(b))`

mbercx commented:
Superseded by https://github.com/aiidateam/aiida-codtools/pull/117
